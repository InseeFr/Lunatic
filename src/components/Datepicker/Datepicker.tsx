import { useState } from 'react';
import { DatepickerField } from './DatepickerField';
import './Datepicker.scss';
import { customizedComponent } from '../shared/HOC/customizedComponent';
import type { LunaticComponentProps } from '../type';
import { Label } from '../shared/Label/Label';
import { ComponentErrors } from '../shared/ComponentErrors/ComponentErrors';
import { Declarations } from '../shared/Declarations/Declarations';

export function LunaticDatepicker({
	disabled,
	readOnly,
	value = '',
	dateFormat = 'YYYY-MM-DD',
	response,
	id,
	handleChange,
	label,
	errors,
	description,
	declarations,
}: LunaticComponentProps<'Datepicker'>) {
	const labelId = `lunatic-datepicker-${id}`;
	const showDay = dateFormat.includes('DD');
	const showMonth = dateFormat.includes('MM');

	// Raw state, we allow invalid dates to be typed
	const [numbers, setNumbers] = useState(() =>
		numbersFromDateString(value ?? undefined)
	);
	const setNumber = (index: number) => (value: number) => {
		const newNumbers = [...numbers] as typeof numbers;
		newNumbers[index] = value;
		setNumbers(newNumbers);
		onChange(newNumbers);
	};

	const onChange = (numbers: [number, number, number]) => {
		const formatParts = dateFormat.split('-');
		const hasNaNIndex = numbers.findIndex((v) => Number.isNaN(v));

		// Date has a missing part
		if (hasNaNIndex > -1 && hasNaNIndex <= formatParts.length - 1) {
			handleChange(response, null);
			return;
		}

		// Date is not valid
		if (dateFormat === 'YYYY-MM-DD' && !isDateValid(numbers)) {
			handleChange(response, null);
			return;
		}

		const result = formatParts
			.map((v, k) => numbers[k].toString().padStart(v.length, '0'))
			.join('-');
		handleChange(response, result);
	};

	const extraProps = {
		readOnly,
		disabled,
	};

	return (
		<div className="lunatic-input">
			<Label htmlFor={id} id={labelId} description={description}>
				{label}
			</Label>
			<Declarations
				type="AFTER_QUESTION_TEXT"
				declarations={declarations}
				id={id}
			/>
			<div className="lunaticDatepickerFields">
				{showDay && (
					<DatepickerField
						id={id + 'day'}
						label="Jour"
						description="Exemple: 14"
						max={31}
						value={numbers[2]}
						onChange={setNumber(2)}
						{...extraProps}
					/>
				)}
				{showMonth && (
					<DatepickerField
						id={id + 'month'}
						label="Mois"
						description="Exemple: 07"
						max={12}
						value={numbers[1]}
						onChange={setNumber(1)}
						{...extraProps}
					/>
				)}
				<DatepickerField
					id={id + 'year'}
					label="AnnÃ©e"
					description="Exemple: 2023"
					value={numbers[0]}
					max={9999}
					onChange={setNumber(0)}
					{...extraProps}
				/>
			</div>
			<ComponentErrors errors={errors} componentId={id} />
		</div>
	);
}

function numbersFromDateString(s?: string): [number, number, number] {
	if (!s) {
		return [NaN, NaN, NaN];
	}
	const parts = s.split('-');
	return [
		parseInt(parts[0], 10),
		parseInt(parts[1], 10),
		parseInt(parts[2], 10),
	];
}

function isDateValid(dateArray: [number, number, number]) {
	const [year, month, day] = dateArray;
	const date = new Date(year, month - 1, day);

	return (
		date.getFullYear() === year &&
		date.getMonth() === month - 1 &&
		date.getDate() === day
	);
}

export const Datepicker = customizedComponent('Datepicker', LunaticDatepicker);
