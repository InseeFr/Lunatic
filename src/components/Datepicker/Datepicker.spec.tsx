import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { Datepicker } from './Datepicker';

describe('Datepicker', () => {
	const mockOnChange = vi.fn();
	const baseProps = {
		value: '1999-01-01',
		executeExpression: vi.fn,
		response: { name: 'demo' },
		'aria-labelledby': 'datepicker',
		id: 'datepicker',
	};

	beforeEach(() => {
		mockOnChange.mockClear();
	});

	(['YYYY-MM-DD', 'YYYY-MM', 'YYYY'] as const).forEach((format) => {
		it('should render properly with format ' + format, () => {
			const { container } = render(
				<Datepicker
					{...baseProps}
					handleChanges={mockOnChange}
					dateFormat={format}
				/>
			);
			expect(container).toMatchSnapshot();
		});
	});

	it('handle change correctly for format YYYY-MM-DD', () => {
		render(
			<Datepicker
				{...baseProps}
				dateFormat="YYYY-MM-DD"
				handleChanges={mockOnChange}
			/>
		);
		fireEvent.change(screen.getByLabelText(/Année/), {
			target: { valueAsNumber: 2023 },
		});
		expect(mockOnChange).toHaveBeenLastCalledWith([
			{
				...baseProps.response,
				value: '2023-01-01',
			},
		]);
		fireEvent.change(screen.getByLabelText(/Mois/), {
			target: { valueAsNumber: 2 },
		});
		fireEvent.change(screen.getByLabelText(/Jour/), {
			target: { valueAsNumber: 30 },
		});
		expect(mockOnChange).toHaveBeenLastCalledWith([
			{ ...baseProps.response, value: null },
		]);
	});

	it('handle change correctly for format YYYY-MM', () => {
		render(
			<Datepicker
				{...baseProps}
				dateFormat="YYYY-MM"
				value="1999-01"
				handleChanges={mockOnChange}
			/>
		);
		fireEvent.change(screen.getByLabelText(/Année/), {
			target: { valueAsNumber: 2023 },
		});
		expect(mockOnChange).toHaveBeenLastCalledWith([
			{
				...baseProps.response,
				value: '2023-01',
			},
		]);
		fireEvent.change(screen.getByLabelText(/Mois/), {
			target: { valueAsNumber: 10 },
		});
		expect(mockOnChange).toHaveBeenLastCalledWith([
			{
				...baseProps.response,
				value: '2023-10',
			},
		]);
	});

	it('handle change correctly for year YYYY', () => {
		render(
			<Datepicker
				{...baseProps}
				dateFormat="YYYY"
				value="1999"
				handleChanges={mockOnChange}
			/>
		);
		fireEvent.change(screen.getByLabelText(/Année/), {
			target: { valueAsNumber: 2023 },
		});
		expect(mockOnChange).toHaveBeenLastCalledWith([
			{ ...baseProps.response, value: '2023' },
		]);
	});
});
