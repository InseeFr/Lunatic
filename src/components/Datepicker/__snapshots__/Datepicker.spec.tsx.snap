// Vitest Snapshot v1

exports[`Datepicker > should render properly with format YYYY 1`] = `
<div>
  <div
    class="lunatic-input"
  >
    <div
      class="lunaticDatepickerFields"
    >
      <div
        class="lunaticDatepickerField lunaticDatepickerFieldLarge"
      >
        <label
          for="datepickeryear"
        >
          Année
          <span
            class="lunaticDatepickerHint"
          >
            Exemple: 2023
          </span>
        </label>
        <input
          id="datepickeryear"
          max="9999"
          min="1"
          type="number"
          value="1999"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Datepicker > should render properly with format YYYY-MM 1`] = `
<div>
  <div
    class="lunatic-input"
  >
    <div
      class="lunaticDatepickerFields"
    >
      <div
        class="lunaticDatepickerField"
      >
        <label
          for="datepickermonth"
        >
          Mois
          <span
            class="lunaticDatepickerHint"
          >
            Exemple: 07
          </span>
        </label>
        <input
          id="datepickermonth"
          max="12"
          min="1"
          type="number"
          value="1"
        />
      </div>
      <div
        class="lunaticDatepickerField lunaticDatepickerFieldLarge"
      >
        <label
          for="datepickeryear"
        >
          Année
          <span
            class="lunaticDatepickerHint"
          >
            Exemple: 2023
          </span>
        </label>
        <input
          id="datepickeryear"
          max="9999"
          min="1"
          type="number"
          value="1999"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Datepicker > should render properly with format YYYY-MM-DD 1`] = `
<div>
  <div
    class="lunatic-input"
  >
    <div
      class="lunaticDatepickerFields"
    >
      <div
        class="lunaticDatepickerField"
      >
        <label
          for="datepickerday"
        >
          Jour
          <span
            class="lunaticDatepickerHint"
          >
            Exemple: 14
          </span>
        </label>
        <input
          id="datepickerday"
          max="31"
          min="1"
          type="number"
          value="1"
        />
      </div>
      <div
        class="lunaticDatepickerField"
      >
        <label
          for="datepickermonth"
        >
          Mois
          <span
            class="lunaticDatepickerHint"
          >
            Exemple: 07
          </span>
        </label>
        <input
          id="datepickermonth"
          max="12"
          min="1"
          type="number"
          value="1"
        />
      </div>
      <div
        class="lunaticDatepickerField lunaticDatepickerFieldLarge"
      >
        <label
          for="datepickeryear"
        >
          Année
          <span
            class="lunaticDatepickerHint"
          >
            Exemple: 2023
          </span>
        </label>
        <input
          id="datepickeryear"
          max="9999"
          min="1"
          type="number"
          value="1999"
        />
      </div>
    </div>
  </div>
</div>
`;
